pisek/task_jobs/builder/strategies.py

```diff
@@ -249,10 +250,31 @@ def applicable_on_files(cls, build: "BuildConfig", sources: list[str]) -> bool:
 
     def _build(self) -> str:
         pas_flags = ["-gl", "-O3", "-Sg", "-o" + self.target, "-FE."]
-        return self._run_compilation(
+        self._run_subprocess(
             ["fpc"] + pas_flags + self.sources + self._build_section.comp_args,
             self._build_section.program_name,
         )
+        return self.target
+
+
+class Make(BuildStrategy):
+    name = BuildStrategyName.make
+
+    @classmethod
+    def applicable_on_files(cls, build: "BuildConfig", sources: list[str]) -> bool:
+        return False
+
+    @classmethod
+    def applicable_on_directory(cls, build: "BuildConfig", directory: str) -> bool:
+        return os.path.exists(os.path.join(directory, "Makefile"))
+
+    def _build(self) -> str:
+        directory = os.listdir()[0]
+        with ChangedCWD(directory):
+            self._run_subprocess(["make"], self._build_section.program_name)
+            if not os.path.isdir("out"):
```

Comment:
Why did you choose `out` as the directory name? I don't think I've ever seen a build directory named like that. Plus, I don't think the name makes much sense. `out` implies it contains outputs of the program, but it doesn't - It contains the runnable program itself. I suppose it contains the output of the compiler, but by the same logic every directory or file not created by hand could be called `out`. What about `build` or `bin`?

Reply:
I would expect that `build` contains temporary files generated during building (e.g., object files), not necessarily the final output. `bin` implies executable files, but the output of the build process is not limited to executables (e.g., Python libraries usually aren't).

During building, `out` is really the output of the build process. When the built program is executed later, the directory is no longer called that.

If you have a better name, tell us please. But I don't think that either `build` or `bin` is better.

Reply:
Sure, `build` does often also contain artifacts, but I think it's pretty common for it to (also) contain the build target. `bin` also makes sense - the output of any Pisek build will be one or more executables. Sure, sometimes there will be supporting files, but those can also be in the `bin` folder. It doesn't match the FSH definition of the folder, but the Dotnet build system does that.

Sure, Pisek renames and moves the directory before executing the binary, but during development, you may very well want to run build and run the command manually, in which case calling the directory `out` doesn't make much sense. I think it makes sense to pick a meaningfull name for when the program is ran outside of Pisek, since when it's built by Pisek, the name is hidden from the user anyway.

Reply:
Well, I wouldn't draw inspiration from the Dotnet build system, since Dotnet is quite alien to the UNIX world. (In fact, when I see Dotnet doing something, I immediately think if doing the _opposite_ is the right choice ;)) Besides Dotnet, I don't remember seeing any serious build system putting non-executable things into `bin`. It's not FHS, it's UNIX tradition that goes much deeper.

Anyway, while I don't consider `build` to be a good choice, it's certainly an acceptable one.

Reply:
Re running things outside Pisek: I don't think that there is any difference between `out` and `build` - the user has to remember that they need to refer to a subdirectory and what is the subdirectory's name.

Reply:
I like `target`.